mlir/lib/Dialect/Rock/utility/loweringUtils.cpp

```diff
@@ -132,18 +133,42 @@ mlir::rock::backwardDataKernelIds(int64_t strideHeight, int64_t strideWidth,
 
   // Populate the kernel IDs according to the current backward data convolution
   // algorithm implementation.
-  for (int64_t kernelId = 0; kernelId < yTilda * xTilda; ++kernelId) {
+  int64_t subproduct = 1;
+  int64_t product;
+  for (size_t i = 1; i < filterDims.size(); i++)
+    subproduct *= filTilda[i];
+  product = subproduct * filTilda[0];
+  for (int64_t kernelId = 0; kernelId < product; ++kernelId) {
     // gemmK size is different for each GEMM
-    const int64_t iYTilda = kernelId / xTilda;
-    const int64_t iXTilda = kernelId % xTilda;
+    SmallVector<int64_t, 3> iTilda;
+    SmallVector<int64_t, 3> iDotSlice;
+    int64_t divisor = 1;
+    iTilda.resize(filterDims.size());
+    switch (filterDims.size()) {
+    default:
+      llvm_unreachable("Only 2-D and 3-D have been implemented.");
+      break;
+    case 3:
+      divisor = filTilda[2];
+      iTilda[2] = kernelId % divisor;
+      [[fallthrough]];
+    case 2:
+      iTilda[1] = (kernelId % subproduct) / divisor;
+      iTilda[0] = kernelId / subproduct;
```

Comment:
I think this can be generalized to N dimensions. 

... heck, you probably want `delinearize` from `external/llvm-project/mlir/include/mlir/Dialect/Utils/IndexingUtils.h`
, which makes it `iTilda = delinearize(kernelID, filDims);`

Reply:
The delinearize() function is computing something slightly different, which is giving me different results in practice.  Could be I need to call it differently, or could be that our needs just differ.

Reply:
Could be, but yeah, this is a loop

```
SmallVector<int64_t> iTilda(0, filterDims.size());
int64_t remainingKernelId = kernelId;
for (auto [thisITilda, dim] : llvm::zip_equal(llvm::reverse(iTilda, filterDims)) {
  thisITilda = remainingKernelId % dim;
  remainingKernelId /= dim;
}
```

Reply:
Which should be exactly `delinearize(kernelId, filterDims);`